frame/im-online/src/lib.rs

```diff
@@ -246,6 +252,9 @@ pub trait Trait: SendTransactionTypes<Call<Self>> + pallet_session::historical::
 	/// there is a chance the authority will produce a block and they won't be necessary.
 	type SessionDuration: Get<Self::BlockNumber>;
 
+	/// A set of validators expected to be online.
+	type ValidatorSet: ValidatorSet<<Self as pallet_session::Trait>::ValidatorId>;
```

Comment:
I feel the decoupling is incomplete. I'd expect `pallet_session` to be removed from the deps of `im-online` completely, but it's still here. What we got now:
1. Increased boilerplate in the `node/runtime` file
2. Partial decoupling, which may end up either breaking assumptions in the `im-online` pallet (since we still rely on some of the session stuff) or may cause incorrect implementation.

Reply:
>Partial decoupling, which may end up either breaking assumptions in the im-online pallet (since we still rely on some of the session stuff) or may cause incorrect implementation.

Yeah, also had this concern, thus want to get some early feedback.

Removing the `pallet_session` dep from `im-online` completely will take more works, but I can try.

Reply:
@tomusdrw After taking a quick at the code, I realize that this decoupling requires quite a refactoring, any guidance on this task? Just in case of I go too far in the wrong direction.

Reply:
@tomusdrw Can you take another look my latest commit? Still have some works undone, but want to ensure I'm on the right track.